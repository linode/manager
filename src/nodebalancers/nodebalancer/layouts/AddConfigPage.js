import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';

import { getObjectByLabelLazily, objectFromMapByLabel } from '~/api/util';
import { nodebalancers } from '~/api';

import { reduceErrors } from '~/errors';
import { Card } from '~/components/cards';
import { setError } from '~/actions/errors';
import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import { ConfigForm } from '../components/ConfigForm';

export class AddConfigPage extends Component {
  static async preload({ dispatch, getState }, { nbLabel }) {
    try {
      const { id } = await dispatch(getObjectByLabelLazily('nodebalancers', nbLabel));
      await dispatch(nodebalancers.configs.all([id]));
    } catch (response) {
      // eslint-disable-next-line no-console
      console.error(response);
      dispatch(setError(response));
    }
  }

  constructor(props) {
    super(props);
    this.saveChanges = this.saveChanges.bind(this);
    this.state = {
      loading: false,
      errors: {},
    };
  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle('Nodebalancers'));
  }

  async saveChanges(stateValues) {
    const { dispatch, nbLabel, nodebalancer } = this.props;

    const {
      port,
      protocol,
      algorithm,
      stickiness,
      check,
      checkPassive,
      checkInterval,
      checkTimeout,
      checkAttempts,
    } = stateValues;
    this.setState({ loading: true, errors: {} });
    const data = {
      port: parseInt(port),
      protocol,
      algorithm,
      stickiness,
      check,
      check_passive: Number(checkPassive),
      check_interval: checkInterval,
      check_timeout: checkTimeout,
      check_attempts: checkAttempts,
    };
    try {
      await dispatch(nodebalancers.configs.post(data, nodebalancer.id));
      this.setState({ loading: false });
      dispatch(push(`/nodebalancers/${nbLabel}`));
    } catch (response) {
      const errors = await reduceErrors(response);
      this.setState({ errors, loading: false });
    }
  }

  render() {
    const { nbLabel } = this.props;
    const { loading, errors } = this.state;

    return (
      <div>
        <header className="main-header main-header--border">
          <div className="container">
            <h1>
              {nbLabel}
            </h1>
          </div>
        </header>
        <div className="container">
          <Card title="Add Configuration">
            <div>
              <p>
                Configure how your NodeBalancer listens for incoming traffic
                and connects to backend nodes. <a href="#">Learn more.</a>
              </p>
            </div>
            <ConfigForm
              saveChanges={this.saveChanges}
              loading={loading}
              errors={errors}
              submitText="Add Configuration"
            />
          </Card>
        </div>
      </div>
    );
  }
}

AddConfigPage.propTypes = {
  dispatch: PropTypes.func,
  nbLabel: PropTypes.string,
  nodebalancer: PropTypes.object,
};

function select(state, ownProps) {
  const params = ownProps.params;
  const nbLabel = params.nbLabel;

  const nodebalancer = objectFromMapByLabel(state.api.nodebalancers.nodebalancers, nbLabel);

  return {
    nbLabel: nbLabel,
    nodebalancer: nodebalancer,
  };
}

export default connect(select)(AddConfigPage);

