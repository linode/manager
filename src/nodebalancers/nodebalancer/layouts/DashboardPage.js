import _ from 'lodash';
import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';

import { Card, CardHeader } from 'linode-components/cards';
import { Table } from 'linode-components/tables';
import { List } from 'linode-components/lists';
import { ListBody } from 'linode-components/lists/bodies';
import { LinkCell, ButtonCell } from 'linode-components/tables/cells';
import { Select } from 'linode-components/forms';
import { DeleteModalBody } from 'linode-components/modals';

import { setSource } from '~/actions/source';
import { showModal, hideModal } from '~/actions/modal';
import { objectFromMapByLabel, getObjectByLabelLazily } from '~/api/util';
import { nodebalancerStats } from '~/api/nodebalancers';
import { nodebalancers } from '~/api';
import Region from '~/linodes/components/Region';
import { dispatchOrStoreErrors } from '~/api/util';
import LineGraph from '~/components/graphs/LineGraph';
import {
  NODEBALANCER_CONFIG_ALGORITHMS, NODEBALANCER_CONFIG_STICKINESS,
} from '~/constants';


const CONNECTION_UNITS = [' connections', 'K connections', 'M connections'];
const NETWORK_UNITS = [' bits', 'K bits', 'M bits'];

function formatData(colors, datasets, legends) {
  const x = datasets[0].map(([x]) => x);
  const ys = datasets.map(dataset => dataset.map(([, y]) => y));
  return LineGraph.formatData(x, ys, colors, legends);
}

export class DashboardPage extends Component {
  static async preload({ dispatch, getState }, { nodebalancerLabel }) {
    const { id } = await dispatch(getObjectByLabelLazily('nodebalancers', nodebalancerLabel));

    try {
      await dispatch(nodebalancerStats([id]));
    } catch (e) {
      // Stats aren't available.
    }
  }

  constructor(props) {
    super(props);

    this.state = {
      source: 'connections',
      units: 0,
    };

    this.componentWillReceiveProps = this.componentWillMount;
  }

  componentWillMount() {

  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));
  }

  shouldComponentUpdate(newProps, newState) {
    // Prevents graph animation from happening multiple times for unchanged data.
    return !_.isEqual(this.props, newProps) || !_.isEqual(this.state, newState);
  }

  onChange = ({ target: { name, value } }) => this.setState({ [name]: value })

  deleteNodeBalancerConfig(nodebalancer, config) {
    const { dispatch } = this.props;
    const title = 'Delete NodeBalancer Config';

    dispatch(showModal(title,
      <DeleteModalBody
        onOk={() => {
          const ids = [nodebalancer.id, config.id].filter(Boolean);

          return dispatch(dispatchOrStoreErrors.call(this, [
            () => nodebalancers.configs.delete(...ids),
            hideModal,
          ]));
        }}
        typeOfItem={title}
        onCancel={() => dispatch(hideModal())}
        items={[`port ${config.port}`]}
      />
    ));
  }

  renderUnitSelect() {
    const { units, source } = this.state;

    const _units = source === 'connections' ? CONNECTION_UNITS : NETWORK_UNITS;

    return (
      <div className="Menu-item">
        <label className="row-label">Units:</label>
        <Select
          value={units}
          name="units"
          onChange={this.onChange}
          options={_units.map((label, value) => ({ label, value }))}
        />
      </div>
    );
  }

  renderGraphs() {
    const { timezone, nodebalancer: { _stats: stats } } = this.props;
    const { units } = this.state;

    if (stats) {
      this.graphs = {
        connections: {
          title: 'Connections',
          yAxis: {
            label: `${CONNECTION_UNITS[units]} per second`,
            format: r => `${r.toFixed(1) / Math.pow(1000, units)}${CONNECTION_UNITS[units]}/s`,
          },
          data: formatData(['990066'], [stats.connections]),
          unit: `${CONNECTION_UNITS[units]}/s`,
        },
        traffic: {
          title: 'Traffic',
          yAxis: {
            label: `${NETWORK_UNITS[units]} per second`,
            format: r => `${r.toFixed(1) / Math.pow(1000, units)}${NETWORK_UNITS[units]}/s`,
          },
          data: formatData(['0033CC', '32CD32'],
                           [stats.traffic.in, stats.traffic.out],
                           ['In', 'Out']),
          unit: `${NETWORK_UNITS[units]}/s`,
        },
      };
    }

    return (
      <Card header={<CardHeader title="Graphs" />}>
        {!this.graphs ? <p>No graphs are available.</p> : (
          <div>
            <div className="Menu">
              <div className="Menu-item">
                <Select
                  value={this.state.source}
                  name="source"
                  onChange={this.onChange}
                >
                  <option value="connections">Connections</option>
                  <option value="traffic">Traffic</option>
                </Select>
              </div>
              {this.renderUnitSelect()}
              <div className="Menu-item Menu-item--right">Last 24 Hours</div>
            </div>
            <LineGraph
              timezone={timezone}
              {...this.graphs[this.state.source]}
            />
          </div>
        )}
      </Card>
    );
  }

  render() {
    const { nodebalancer } = this.props;
    const { configs } = nodebalancer._configs;

    const newConfigs = Object.values(configs).map((config) => {
      return {
        ...config,
        protocol: config.protocol.toUpperCase(),
        algorithm: NODEBALANCER_CONFIG_ALGORITHMS.get(config.algorithm),
        stickiness: NODEBALANCER_CONFIG_STICKINESS.get(config.stickiness),
        statusString: `${config.nodes_status.up} up, ${config.nodes_status.down} down`,
      };
    });

    return (
      <div>
        <section>
          <Card header={<CardHeader title="Summary" />}>
            <div className="row">
              <div className="col-sm-2 row-label">
                IP Addresses
              </div>
              <div className="col-sm-10">
                <ul className="list-unstyled">
                  <li>{nodebalancer.ipv4}</li>
                  <li className="text-muted">{nodebalancer.ipv6}</li>
                </ul>
              </div>
            </div>
            <div className="row">
              <div className="col-sm-2 row-label">
                Hostname
              </div>
              <div className="col-sm-10">
                {nodebalancer.hostname}
              </div>
            </div>
            <div className="row">
              <div className="col-sm-2 row-label">
                Region
              </div>
              <div className="col-sm-10">
                <Region obj={nodebalancer} />
              </div>
            </div>
          </Card>
        </section>
        <section>
          <Card
            header={
              <CardHeader
                title="Configs"
                nav={
                  <Link
                    to={`/nodebalancers/${nodebalancer.label}/configs/create`}
                    className="linode-add btn btn-default float-sm-right"
                  >Add a Config</Link>
                }
              />
            }
          >
            <List>
              <ListBody>
                <Table
                  className="Table--secondary"
                  columns={[
                    { textKey: 'port', label: 'Port',
                      cellComponent: LinkCell,
                      hrefFn: function (config) {
                        return `/nodebalancers/${nodebalancer.label}/configs/${config.id}`;
                      },
                      tooltipEnabled: true,
                    },
                    { dataKey: 'protocol', label: 'Protocol' },
                    { dataKey: 'algorithm', label: 'Algorithm' },
                    { dataKey: 'stickiness', label: 'Session Stickiness' },
                    { dataKey: 'statusString', label: 'Node Status' },
                    {
                      cellComponent: ButtonCell,
                      headerClassName: 'ButtonColumn',
                      onClick: (config) => {
                        this.deleteNodeBalancerConfig(nodebalancer, config);
                      },
                      text: 'Delete',
                    },
                  ]}
                  data={newConfigs}
                  noDataMessage="You have no configs."
                  selectedMap={{}}
                />
              </ListBody>
            </List>
          </Card>
        </section>
        {this.renderGraphs()}
      </div>
    );
  }
}

DashboardPage.propTypes = {
  dispatch: PropTypes.func,
  nodebalancer: PropTypes.object,
  timezone: PropTypes.string,
};

function select(state, ownProps) {
  const params = ownProps.params;
  const nbLabel = params.nbLabel;
  const { timezone } = state.api.profile;

  const nodebalancer = objectFromMapByLabel(state.api.nodebalancers.nodebalancers, nbLabel);

  return { nodebalancer, timezone };
}

export default connect(select)(DashboardPage);
