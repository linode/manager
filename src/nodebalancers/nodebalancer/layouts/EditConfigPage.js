import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { push } from 'react-router-redux';

import { getObjectByLabelLazily, objectFromMapByLabel } from '~/api/util';
import { nodebalancers } from '~/api';

import { reduceErrors } from '~/errors';
import { Card, CardHeader } from '~/components/cards';
import { setError } from '~/actions/errors';
import { setSource } from '~/actions/source';
import { setTitle } from '~/actions/title';
import { ConfigForm } from '../components/ConfigForm';

export class EditConfigPage extends Component {
  static async preload({ dispatch, getState }, { nbLabel }) {
    try {
      const { id } = await dispatch(getObjectByLabelLazily('nodebalancers', nbLabel));
      await dispatch(nodebalancers.configs.all([id]));
    } catch (response) {
      // eslint-disable-next-line no-console
      console.error(response);
      await dispatch(setError(response));
    }
  }

  constructor(props) {
    super(props);
    this.saveChanges = this.saveChanges.bind(this);
    this.state = {
      loading: false,
      errors: {},
    };
  }

  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));

    dispatch(setTitle('Nodebalancers'));
  }

  async saveChanges(stateValues) {
    const { config, dispatch, nbLabel, nodebalancer } = this.props;

    const {
      port,
      protocol,
      algorithm,
      stickiness,
      check,
      checkPassive,
      checkInterval,
      checkTimeout,
      checkAttempts,
    } = stateValues;
    this.setState({ loading: true, errors: {} });
    const data = {
      port: parseInt(port),
      protocol,
      algorithm,
      stickiness,
      check,
      check_passive: Number(checkPassive),
      check_interval: checkInterval,
      check_timeout: checkTimeout,
      check_attempts: checkAttempts,
    };
    try {
      await dispatch(nodebalancers.configs.put(data, nodebalancer.id, config.id));
      this.setState({ loading: false });
      await dispatch(push(`/nodebalancers/${nbLabel}`));
    } catch (response) {
      const errors = await reduceErrors(response);
      this.setState({ errors, loading: false });
    }
  }

  render() {
    const { config, nbLabel, dispatch } = this.props;
    const { loading, errors } = this.state;

    if (!config) {
      dispatch(push(`/nodebalancers/${nbLabel}`));
      return null;
    }

    return (
      <div>
        <header className="main-header main-header--border">
          <div className="container">
            <h1>
              {nbLabel}
            </h1>
          </div>
        </header>
        <div className="container">
          <Card header={<CardHeader title="Edit Configuration" />}>
            <div>
              <p>
                {/* eslint-disable max-len */}
                Configure how your NodeBalancer listens for incoming traffic
                and connects to backend nodes. <a href="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer" target="_blank">Learn more.</a>
                {/* eslint-enable max-len */}
              </p>
            </div>
            <ConfigForm
              saveChanges={this.saveChanges}
              loading={loading}
              errors={errors}
              submitText="Edit Configuration"
              port={config.port}
              protocol={config.protocol}
              algorithm={config.algorithm}
              stickiness={config.stickiness}
              check={config.check}
              checkPassive={!!config.check_passive}
              checkInterval={config.check_interval}
              checkTimeout={config.check_timeout}
              checkAttempts={config.check_attempts}
              nodebalancerConfigId={config.id}
            />
          </Card>
        </div>
      </div>
    );
  }
}

EditConfigPage.propTypes = {
  dispatch: PropTypes.func,
  config: PropTypes.object,
  nbLabel: PropTypes.string,
  nodebalancer: PropTypes.object,
  port: PropTypes.number,
};

function select(state, ownProps) {
  const params = ownProps.params;
  const nbLabel = params.nbLabel;
  const id = parseInt(params.port);

  const nodebalancer = objectFromMapByLabel(state.api.nodebalancers.nodebalancers, nbLabel);
  const config = objectFromMapByLabel(nodebalancer._configs.configs, id, 'id');

  return {
    nbLabel: nbLabel,
    id: id,
    nodebalancer: nodebalancer,
    config: config,
  };
}
export default connect(select)(EditConfigPage);
