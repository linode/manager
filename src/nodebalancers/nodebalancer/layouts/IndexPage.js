import React, { Component, PropTypes } from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { push } from 'react-router-redux';
import { Tabs } from 'linode-components/tabs';

import Breadcrumbs from '~/components/Breadcrumbs';

import { nodebalancers } from '~/api';
import { getObjectByLabelLazily, objectFromMapByLabel } from '~/api/util';
import { setTitle } from '~/actions/title';
import { GroupLabel } from '~/components';


export class IndexPage extends Component {
  static async preload({ dispatch, getState }, { nbLabel }) {
    const { id } = await dispatch(getObjectByLabelLazily('nodebalancers', nbLabel));
    await dispatch(nodebalancers.configs.all([id]));
  }

  async componentDidMount() {
    const { dispatch, nodebalancer } = this.props;
    dispatch(setTitle(nodebalancer.label));
  }

  render() {
    const { nodebalancer } = this.props;
    if (!nodebalancer) return null;

    const { label } = nodebalancer;
    const tabs = [
      { name: 'Dashboard', link: `/nodebalancers/${label}` },
      { name: 'Settings', link: `/nodebalancers/${label}/settings` },
    ];

    return (
      <div>
        <header className="main-header">
          <div className="container clearfix">
            <div className="float-sm-left">
              <Breadcrumbs crumbs={[{ to: '/nodebalancers', label: 'NodeBalancers' }]} />
              <h1 title={nodebalancer.id}>
                <Link to={`/nodebalancers/${nodebalancer.label}`}>
                  <GroupLabel object={nodebalancer} />
                </Link>
              </h1>
            </div>
          </div>
        </header>
        <div className="main-header-fix"></div>
        <Tabs
          tabs={tabs}
          onClick={(e, tabIndex) => {
            e.stopPropagation();
            this.props.dispatch(push(tabs[tabIndex].link));
          }}
          pathname={location.pathname}
        >
          {this.props.children}
        </Tabs>
      </div>
    );
  }
}

IndexPage.propTypes = {
  dispatch: PropTypes.func,
  nodebalancer: PropTypes.object,
  children: PropTypes.node,
};

function select(state, props) {
  const { nodebalancers } = state.api.nodebalancers;
  const { nbLabel } = props.params;
  const nodebalancer = objectFromMapByLabel(Object.values(nodebalancers), nbLabel);
  return { nodebalancer };
}

export default connect(select)(IndexPage);
