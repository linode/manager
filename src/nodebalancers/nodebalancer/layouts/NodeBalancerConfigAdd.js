import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { ExternalLink } from 'linode-components';
import { Card, CardHeader } from 'linode-components';

import { setSource } from '~/actions/source';
import { objectFromMapByLabel } from '~/api/util';

import { ConfigForm } from '../components';


export class NodeBalancerConfigAdd extends Component {
  async componentDidMount() {
    const { dispatch } = this.props;
    dispatch(setSource(__filename));
  }

  render() {
    const { dispatch, nodebalancer } = this.props;

    return (
      <Card header={<CardHeader title="Add Config" />}>
        <p>
          {/* eslint-disable max-len */}
          Configure how your NodeBalancer listens for incoming traffic
          and connects to backend nodes. <ExternalLink to="https://www.linode.com/docs/platform/nodebalancer/getting-started-with-nodebalancers#configuring-a-nodebalancer">Learn more.</ExternalLink>
          {/* eslint-enable max-len */}
        </p>
        <ConfigForm
          dispatch={dispatch}
          nodebalancer={nodebalancer}
          submitText="Add Config"
          submitDisabledText="Adding Config"
        />
      </Card>
    );
  }
}

NodeBalancerConfigAdd.propTypes = {
  dispatch: PropTypes.func,
  nodebalancer: PropTypes.object,
};

function mapStateToProps(state, ownProps) {
  const params = ownProps.match.params;
  const nbLabel = params.nbLabel;

  const nodebalancer = objectFromMapByLabel(state.api.nodebalancers.nodebalancers, nbLabel);

  return { nodebalancer };
}

export default connect(mapStateToProps)(NodeBalancerConfigAdd);

