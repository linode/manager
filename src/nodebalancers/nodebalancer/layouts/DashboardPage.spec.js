import React from 'react';
import sinon from 'sinon';
import { mount, shallow } from 'enzyme';
import { StaticRouter } from 'react-router-dom';

import { configsNodeBalancer } from '~/data/nodebalancers';
import { DashboardPage } from '~/nodebalancers/nodebalancer/layouts/DashboardPage';

describe('nodebalancers/nodebalancer/layouts/DashboardPage', () => {
  const sandbox = sinon.sandbox.create();

  afterEach(() => {
    sandbox.restore();
  });

  const dispatch = sandbox.stub();

  it('should render without error', () => {
    const dispatch = jest.fn();
    const wrapper = shallow(
      <StaticRouter>
        <DashboardPage
          dispatch={dispatch}
          nodebalancer={configsNodeBalancer}
          transfer={{ used: 1, quota: 5 }}
        />
      </StaticRouter>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('displays nodebalancer configs', () => {
    const page = mount(
      <StaticRouter>
        <DashboardPage
          dispatch={dispatch}
          nodebalancer={configsNodeBalancer}
          transfer={{ used: 1, quota: 5 }}
        />
      </StaticRouter>
    );
    const firstRow = page.find('tr').at(1);
    const secondRow = page.find('tr').at(2);
    const renderedPort = firstRow.find('td').at(0).text();
    const renderedProtocol = firstRow.find('td').at(1).text();
    const renderedAlgorithm = firstRow.find('td').at(2).text();
    const renderedStick = firstRow.find('td').at(3).text();
    const renderedCheck = firstRow.find('td').at(4).text();
    const secondRenderedPort = secondRow.find('td').at(0).text();
    expect(renderedPort).toBe('80');
    expect(renderedProtocol).toBe('HTTP');
    expect(renderedAlgorithm).toBe('Round Robin');
    expect(renderedStick).toBe('-- None --');
    expect(renderedCheck).toBe('1 up, 0 down');
    expect(secondRenderedPort).toBe('81');
  });
});

