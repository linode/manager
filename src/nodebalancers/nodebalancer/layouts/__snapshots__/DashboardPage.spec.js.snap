// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nodebalancers/nodebalancer/layouts/DashboardPage should render without error 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <StaticRouter
    basename=""
    location="/"
  >
    <DashboardPage
      dispatch={[Function]}
      nodebalancer={
        Object {
          "_configs": Object {
            "configs": Object {
              "1": Object {
                "_nodes": Object {
                  "nodes": Object {
                    "1": Object {
                      "address": "192.168.4.5:80",
                      "config_id": 0,
                      "id": 1,
                      "label": "greatest_node_ever",
                      "mode": "accept",
                      "nodebalancer_id": 23,
                      "status": "online",
                      "weight": 40,
                    },
                  },
                },
                "algorithm": "roundrobin",
                "check": "none",
                "check_attempts": 3,
                "check_body": "",
                "check_interval": 0,
                "check_passive": true,
                "check_path": "",
                "check_timeout": 30,
                "cipher_suite": "recommended",
                "id": 1,
                "nodebalancer_id": 23,
                "nodes_status": Object {
                  "down": 0,
                  "up": 1,
                },
                "port": 80,
                "protocol": "http",
                "stickiness": "none",
              },
              "2": Object {
                "algorithm": "roundrobin",
                "check": "none",
                "check_attempts": 3,
                "check_body": "",
                "check_interval": 0,
                "check_passive": true,
                "check_path": "",
                "check_timeout": 30,
                "cipher_suite": "recommended",
                "id": 2,
                "nodebalancer_id": 23,
                "nodes_status": Object {
                  "down": 0,
                  "up": 0,
                },
                "port": 81,
                "protocol": "http",
                "stickiness": "none",
              },
            },
          },
          "client_conn_throttle": 0,
          "created": "2017-03-03T18:08:44",
          "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
          "id": 23,
          "ipv4": "1.1.1.1",
          "label": "nodebalancer-1",
          "region": "us-east-1a",
          "status": "new_active",
          "updated": "2017-03-06T22:00:48",
        }
      }
      transfer={
        Object {
          "quota": 5,
          "used": 1,
        }
      }
    />
  </StaticRouter>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <DashboardPage
        dispatch={[Function]}
        nodebalancer={
          Object {
            "_configs": Object {
              "configs": Object {
                "1": Object {
                  "_nodes": Object {
                    "nodes": Object {
                      "1": Object {
                        "address": "192.168.4.5:80",
                        "config_id": 0,
                        "id": 1,
                        "label": "greatest_node_ever",
                        "mode": "accept",
                        "nodebalancer_id": 23,
                        "status": "online",
                        "weight": 40,
                      },
                    },
                  },
                  "algorithm": "roundrobin",
                  "check": "none",
                  "check_attempts": 3,
                  "check_body": "",
                  "check_interval": 0,
                  "check_passive": true,
                  "check_path": "",
                  "check_timeout": 30,
                  "cipher_suite": "recommended",
                  "id": 1,
                  "nodebalancer_id": 23,
                  "nodes_status": Object {
                    "down": 0,
                    "up": 1,
                  },
                  "port": 80,
                  "protocol": "http",
                  "stickiness": "none",
                },
                "2": Object {
                  "algorithm": "roundrobin",
                  "check": "none",
                  "check_attempts": 3,
                  "check_body": "",
                  "check_interval": 0,
                  "check_passive": true,
                  "check_path": "",
                  "check_timeout": 30,
                  "cipher_suite": "recommended",
                  "id": 2,
                  "nodebalancer_id": 23,
                  "nodes_status": Object {
                    "down": 0,
                    "up": 0,
                  },
                  "port": 81,
                  "protocol": "http",
                  "stickiness": "none",
                },
              },
            },
            "client_conn_throttle": 0,
            "created": "2017-03-03T18:08:44",
            "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
            "id": 23,
            "ipv4": "1.1.1.1",
            "label": "nodebalancer-1",
            "region": "us-east-1a",
            "status": "new_active",
            "updated": "2017-03-06T22:00:48",
          }
        }
        transfer={
          Object {
            "quota": 5,
            "used": 1,
          }
        }
      />,
      "history": Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "dispatch": [Function],
        "nodebalancer": Object {
          "_configs": Object {
            "configs": Object {
              "1": Object {
                "_nodes": Object {
                  "nodes": Object {
                    "1": Object {
                      "address": "192.168.4.5:80",
                      "config_id": 0,
                      "id": 1,
                      "label": "greatest_node_ever",
                      "mode": "accept",
                      "nodebalancer_id": 23,
                      "status": "online",
                      "weight": 40,
                    },
                  },
                },
                "algorithm": "roundrobin",
                "check": "none",
                "check_attempts": 3,
                "check_body": "",
                "check_interval": 0,
                "check_passive": true,
                "check_path": "",
                "check_timeout": 30,
                "cipher_suite": "recommended",
                "id": 1,
                "nodebalancer_id": 23,
                "nodes_status": Object {
                  "down": 0,
                  "up": 1,
                },
                "port": 80,
                "protocol": "http",
                "stickiness": "none",
              },
              "2": Object {
                "algorithm": "roundrobin",
                "check": "none",
                "check_attempts": 3,
                "check_body": "",
                "check_interval": 0,
                "check_passive": true,
                "check_path": "",
                "check_timeout": 30,
                "cipher_suite": "recommended",
                "id": 2,
                "nodebalancer_id": 23,
                "nodes_status": Object {
                  "down": 0,
                  "up": 0,
                },
                "port": 81,
                "protocol": "http",
                "stickiness": "none",
              },
            },
          },
          "client_conn_throttle": 0,
          "created": "2017-03-03T18:08:44",
          "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
          "id": 23,
          "ipv4": "1.1.1.1",
          "label": "nodebalancer-1",
          "region": "us-east-1a",
          "status": "new_active",
          "updated": "2017-03-06T22:00:48",
        },
        "transfer": Object {
          "quota": 5,
          "used": 1,
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <DashboardPage
          dispatch={[Function]}
          nodebalancer={
            Object {
              "_configs": Object {
                "configs": Object {
                  "1": Object {
                    "_nodes": Object {
                      "nodes": Object {
                        "1": Object {
                          "address": "192.168.4.5:80",
                          "config_id": 0,
                          "id": 1,
                          "label": "greatest_node_ever",
                          "mode": "accept",
                          "nodebalancer_id": 23,
                          "status": "online",
                          "weight": 40,
                        },
                      },
                    },
                    "algorithm": "roundrobin",
                    "check": "none",
                    "check_attempts": 3,
                    "check_body": "",
                    "check_interval": 0,
                    "check_passive": true,
                    "check_path": "",
                    "check_timeout": 30,
                    "cipher_suite": "recommended",
                    "id": 1,
                    "nodebalancer_id": 23,
                    "nodes_status": Object {
                      "down": 0,
                      "up": 1,
                    },
                    "port": 80,
                    "protocol": "http",
                    "stickiness": "none",
                  },
                  "2": Object {
                    "algorithm": "roundrobin",
                    "check": "none",
                    "check_attempts": 3,
                    "check_body": "",
                    "check_interval": 0,
                    "check_passive": true,
                    "check_path": "",
                    "check_timeout": 30,
                    "cipher_suite": "recommended",
                    "id": 2,
                    "nodebalancer_id": 23,
                    "nodes_status": Object {
                      "down": 0,
                      "up": 0,
                    },
                    "port": 81,
                    "protocol": "http",
                    "stickiness": "none",
                  },
                },
              },
              "client_conn_throttle": 0,
              "created": "2017-03-03T18:08:44",
              "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
              "id": 23,
              "ipv4": "1.1.1.1",
              "label": "nodebalancer-1",
              "region": "us-east-1a",
              "status": "new_active",
              "updated": "2017-03-06T22:00:48",
            }
          }
          transfer={
            Object {
              "quota": 5,
              "used": 1,
            }
          }
        />,
        "history": Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
          },
          "push": [Function],
          "replace": [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "dispatch": [Function],
          "nodebalancer": Object {
            "_configs": Object {
              "configs": Object {
                "1": Object {
                  "_nodes": Object {
                    "nodes": Object {
                      "1": Object {
                        "address": "192.168.4.5:80",
                        "config_id": 0,
                        "id": 1,
                        "label": "greatest_node_ever",
                        "mode": "accept",
                        "nodebalancer_id": 23,
                        "status": "online",
                        "weight": 40,
                      },
                    },
                  },
                  "algorithm": "roundrobin",
                  "check": "none",
                  "check_attempts": 3,
                  "check_body": "",
                  "check_interval": 0,
                  "check_passive": true,
                  "check_path": "",
                  "check_timeout": 30,
                  "cipher_suite": "recommended",
                  "id": 1,
                  "nodebalancer_id": 23,
                  "nodes_status": Object {
                    "down": 0,
                    "up": 1,
                  },
                  "port": 80,
                  "protocol": "http",
                  "stickiness": "none",
                },
                "2": Object {
                  "algorithm": "roundrobin",
                  "check": "none",
                  "check_attempts": 3,
                  "check_body": "",
                  "check_interval": 0,
                  "check_passive": true,
                  "check_path": "",
                  "check_timeout": 30,
                  "cipher_suite": "recommended",
                  "id": 2,
                  "nodebalancer_id": 23,
                  "nodes_status": Object {
                    "down": 0,
                    "up": 0,
                  },
                  "port": 81,
                  "protocol": "http",
                  "stickiness": "none",
                },
              },
            },
            "client_conn_throttle": 0,
            "created": "2017-03-03T18:08:44",
            "hostname": "nb-1-1-1-1.newark.nodebalancer.linode.com",
            "id": 23,
            "ipv4": "1.1.1.1",
            "label": "nodebalancer-1",
            "region": "us-east-1a",
            "status": "new_active",
            "updated": "2017-03-06T22:00:48",
          },
          "transfer": Object {
            "quota": 5,
            "used": 1,
          },
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
